From 7d9c616387c2c753b436b4dfd50a835aba3cc1c0 Mon Sep 17 00:00:00 2001
From: Owen O'Malley <omalley@apache.org>
Date: Fri, 4 Mar 2011 04:50:13 +0000
Subject: [PATCH 055/795] MR1: commit 98d8043ef96dd9a67e23852b5e5caf8f0a4589db
 Author: Krishna Ramachandran <ramach@yahoo-inc.com>
 Date:   Thu Oct 21 12:21:50 2010 -0700

    . Delete PrintWriter using iterator to fix
        java.util.ConcurrentModificationException (dking)

git-svn-id: https://svn.apache.org/repos/asf/hadoop/common/branches/branch-0.20-security-patches@1077739 13f79535-47bb-0310-9956-ffa450edef68
(cherry picked from commit dd9742fbca934eaee28c8e8f43e21c99a4b9b51f)
(cherry picked from commit 06d7acc22c74cab6211e9786ed30f4ef6db77561)
---
 .../org/apache/hadoop/mapred/JobHistory.java       |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/mapred/org/apache/hadoop/mapred/JobHistory.java b/src/mapred/org/apache/hadoop/mapred/JobHistory.java
index a644f91..71e6ecf 100644
--- a/src/mapred/org/apache/hadoop/mapred/JobHistory.java
+++ b/src/mapred/org/apache/hadoop/mapred/JobHistory.java
@@ -588,11 +588,12 @@ static void log(ArrayList<PrintWriter> writers, RecordTypes recordType,
     }
     builder.append(LINE_DELIMITER_CHAR);
     
-    for (PrintWriter out : writers) {
+    for (Iterator<PrintWriter> iter = writers.iterator(); iter.hasNext();) {
+      PrintWriter out = iter.next();
       out.println(builder.toString());
       if (out.checkError() && id != null) {
         LOG.info("Logging failed for job " + id + "removing PrintWriter from FileManager");
-        fileManager.removeWriter(id, out);
+        iter.remove();
       }
     }
   }
-- 
1.7.0.4

